From d7387b2adfcac802394d440fc48fc2432d43e17a Mon Sep 17 00:00:00 2001
From: Jerome Robert <jeromerobert@gmx.com>
Date: Sun, 2 Jan 2022 23:37:59 +0100
Subject: [PATCH] Fix compilation with clang 13

---
 src/random_pivot_manager.cpp | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/random_pivot_manager.cpp b/src/random_pivot_manager.cpp
index 4e191e7..f7a30a4 100644
--- a/src/random_pivot_manager.cpp
+++ b/src/random_pivot_manager.cpp
@@ -31,14 +31,16 @@ namespace hmat {
     usedPivots_ +=1;
     if (pivots_.empty())
       return;
-
+#ifndef NDEBUG
     int numberOfPivotsToRemove = 0;
-
+#endif
     for (int i = 0; i < pivots_.size(); ++i) {
       Pivot<dp_t> &pivot = pivots_[i];
       pivot.value_ -= (*row)[pivot.col_] * (*col)[pivot.row_];
+#ifndef NDEBUG
       if (pivot.row_ == rowIndex || pivot.col_ == colIndex)
         numberOfPivotsToRemove += 1;
+#endif
     }
     std::sort(pivots_.begin(), pivots_.end(), Pivot<dp_t>::ComparerLower);
     int size;
-- 
2.25.1

