diff --git a/CMakeLists.txt b/CMakeLists.txt
index 5ad8e50..4b21aec 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -234,10 +234,10 @@ endif ()
 # Warning flags
 # ========================
 include(CheckCXXCompilerFlag)
-check_cxx_compiler_flag("-Werror -Wall -Wno-unused-result -Wno-sign-compare -Wno-literal-suffix" HAVE_GCC_WARNING_FLAGS)
+check_cxx_compiler_flag("-Wall -Wno-unused-result -Wno-sign-compare -Wno-literal-suffix" HAVE_GCC_WARNING_FLAGS)
 if(HAVE_GCC_WARNING_FLAGS)
-  SET(CMAKE_CXX_FLAGS "-Werror -Wall -Wno-unused-result -Wno-sign-compare -Wno-literal-suffix ${CMAKE_CXX_FLAGS}")
-  SET(CMAKE_C_FLAGS "-Werror -Wall -Wno-sign-compare ${CMAKE_C_FLAGS}")
+  SET(CMAKE_CXX_FLAGS "-Wall -Wno-unused-result -Wno-sign-compare -Wno-literal-suffix ${CMAKE_CXX_FLAGS}")
+  SET(CMAKE_C_FLAGS "-Wall -Wno-sign-compare ${CMAKE_C_FLAGS}")
 endif()
 
 # ========================
