diff --git a/CMakeLists.txt b/CMakeLists.txt
index 5ad8e50..1135b63 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -347,6 +347,7 @@ macro(hmat_add_example name source)
     endif ()
     add_executable(${HMAT_PREFIX_EXAMPLE}${name} ${PROJECT_SOURCE_DIR}/examples/${source})
     target_link_libraries(${HMAT_PREFIX_EXAMPLE}${name} ${PROJECT_NAME})
+    target_link_libraries(${HMAT_PREFIX_EXAMPLE}${name} ${LAPACK_LIBRARIES} ${CBLAS_LIBRARIES})
     if (INSTALL_EXAMPLES)
       install(TARGETS ${HMAT_PREFIX_EXAMPLE}${name} DESTINATION "${RELATIVE_INSTALL_BIN_DIR}/examples" COMPONENT Runtime)
     endif ()
